apps/emqx/src/emqx_broker.erl

```diff
@@ -339,28 +342,66 @@ delivery(Msg) -> #delivery{sender = self(), message = Msg}.
 %% Route
 %%--------------------------------------------------------------------
 
--spec route([emqx_types:route_entry()], emqx_types:delivery(), nil() | [persisted]) ->
+route(Routes, Delivery = #delivery{message = _Msg}, PersistRes) ->
+    ?EXT_TRACE_WITH_PROCESS_FUN(
+        msg_route,
+        Delivery,
+        (emqx_otel_trace:msg_attrs(_Msg))#{
+            'route.from' => node(),
+            'route.matched_result' => emqx_utils_json:encode([
+                route_result({TF, RouteTo})
+             || {TF, RouteTo} <- Routes
+            ])
+        },
+        fun(DeliveryWithTrace) ->
+            do_route(Routes, DeliveryWithTrace, PersistRes)
+        end
+    ).
```

Comment:
Suggestions: 

Is there a more simple way?
```
Ctx = emqx_external_trace:start(route),
Output = do_route(Routes, Delivery, PersistRes),
emqx_external_trace:add_attrs(Ctx, ?LAYZ_ATTRS(route, _Input = [Routes, Delivery, PersistRes], Output))
emqx_external_trace:end(Ctx).
```

We only need minimal, straightforward representations to show that tracing was done here, rather than showing the details of how the tracing was done.

We can hide all the details in the `emqx_external_trace.erl` module if possible.

Reply:
I'd prefer this:

```
EXT_TRACE_MSG_ROUTE(Delivery, do_route(Routes, Delivery, PersistRes)).
```

and

```
-define(EXT_TRACE_MSG_ROUTE(Delivery, Expr), 
        %% is traced checks white list, otherwise
        %% the enabl/disable flag, random() >= ratio, and
        case is_traced(Delivery) of
            true ->
                Ctx = emqx_external_trace:start(route),
                Res = Expr,
                ...
                emqx_external_trace:add_attrs(Ctx, ..., Res),
                emqx_external_trace:end(Ctx),
                Res;
            false ->
                Expr
        end
)
```